### ISSUE TITLE
learning from external solution during search

### ISSUE THREAD

**Issue by gregy4** [google/or-tools#1782]
Hello,
  my problem is modeled in a constraint solver and cp-sat solver so I receive valid solutions from both solvers. I see that solvers are weak in different areas. Constraint solver has problem to find better results no matter how much time computation time is given, cp-sat solver is much more sensitive to a horizon (how long schedule can be) so bad estimate can lead to no solution is found or iteration close to optimum takes too much time.

  The problem with a horizon in cp-sat can be for me solved by an event "find only solutions that are shorter than given value" when I try to minimize makespan and execute solvers in parallel. Something that is already done for parallel execution of search workers.

  Is it possible to implement this variant ?  For long computations it makes sense to me to learn new bounds minimally for objective function during search when alternative solver find some solutions faster.

Thanks,
Jan

**Comment by lperron** [comment]
Short answer, you can launch the CP and CP-SAT in parallel, then restart
the CP-SAT model with a new upper bound on the makespan if the CP solver
finds an interesting solution. You can also use this solution as a hint for
the next CP-SAT search.

Long answer, injecting objective bounds is on our roadmap. This is already
part of the protocol between workers, but the API is not ready for prime
time.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00

**Comment by gregy4** [comment]
Thanks Laurent
To clarify it, by restart you mean start cp-sat solver again with new bound ?
I like to answer this kind of questions by myself, but do you have some estimate when injecting of bounds will be available ?  2020 or later ?

**Comment by lperron** [comment]
yes for the first question.

For the timeline, I simply do not know.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00
